src/lib/comms/sol-http-common.c

```diff
@@ -882,3 +883,238 @@ sol_http_split_post_field(const char *query, struct sol_http_params *params)
+    bool any_type1, any_type2;
+
+    pri1 = (const struct sol_http_content_type_priority *)v1;
+    pri2 = (const struct sol_http_content_type_priority *)v2;
+
+    any_type1 = sol_str_slice_str_eq(pri1->type, "*");
+    any_type2 = sol_str_slice_str_eq(pri2->type, "*");
+
+    //text/html have precedence over */* for example.
+    if (!any_type1 && any_type2)
+        return -1;
+    if (any_type1 && !any_type2)
+        return 1;
+
+    //Higher qvalue first
+    if (pri1->qvalue > pri2->qvalue)
+        return -1;
+    if (pri1->qvalue < pri2->qvalue)
+        return 1;
+
+    //Specialized first
+    if (pri1->tokens.len > pri2->tokens.len)
+        return -1;
+    if (pri1->tokens.len < pri2->tokens.len)
+        return 1;
+
+    //Specialized first
+    if (sol_str_slice_eq(pri1->type, pri2->type)) {
+        //text/html have precedence over text/* for example.
+        any_type1 = sol_str_slice_str_eq(pri1->sub_type, "*");
+        any_type2 = sol_str_slice_str_eq(pri2->sub_type, "*");
+
+        if (!any_type1 && any_type2)
+            return -1;
+        if (any_type1 && !any_type2)
+            return 1;
+    }
+
+    //Precedence must prevail
+    if (pri1->index > pri2->index)
+        return 1;
+    if (pri1->index < pri2->index)
+        return -1;
+
+    return 0;
+}
+
+static char *
+is_qvalue_param(const struct sol_str_slice param)
+{
+#define STATE_NEEDS_Q_CHAR (0)
+#define STATE_NEEDS_EQUAL_CHAR (1)
+
+    size_t i;
+    int state = STATE_NEEDS_Q_CHAR;
+
+    for (i = 0; i < param.len; i++) {
+        if (isspace((int)param.data[i]))
+            continue;
+
+        switch (state) {
+        case STATE_NEEDS_Q_CHAR:
+            if (param.data[i] != 'q')
+                return NULL;
+            state = STATE_NEEDS_EQUAL_CHAR;
+            break;
+        default:
+            if (param.data[i] != '=')
+                return NULL;
+            return (char *)param.data + i;
+        }
+    }
+
+    return NULL;
+
+#undef STATE_NEEDS_EQUAL_CHAR
+#undef STATE_NEEDS_Q_CHAR
+}
+
+static int
+set_type_and_sub_type(const struct sol_str_slice content_type,
+    struct sol_str_slice *type, struct sol_str_slice *sub_type)
+{
+    char *sep;
+
+    sep = memchr(content_type.data, '/', content_type.len);
+    SOL_NULL_CHECK(sep, -EINVAL);
+
+    type->data = content_type.data;
+    type->len = sep - content_type.data;
+
+    sub_type->data = sep + 1;
+    sub_type->len = content_type.len - type->len - 1;
+
+    return 0;
+}
+
+SOL_API int
+sol_http_parse_content_type_priorities(const struct sol_str_slice content_type,
+    struct sol_vector *priorities)
+{
+    struct sol_vector tokens;
+    struct sol_str_slice *token;
+    struct sol_http_content_type_priority *pri;
+    uint16_t i;
+
+    SOL_NULL_CHECK(priorities, -EINVAL);
+
+    SOL_DBG("Parsing content priorities for: %.*s",
+        SOL_STR_SLICE_PRINT(content_type));
+    tokens = sol_str_slice_split(content_type, ",", 0);
```

Comment:
hum... it's bad that we need to allocate memory to just iterate. Also, below you do a manual iteration pattern as well (over `';'`), then my recommendation is to follow json-scanner pattern and offer a way to iterate without allocating memory, something like:

```c
struct sol_str_slice type = SOL_STR_SLICE_EMPTY; // first iteration takes an empty string, it is also used to keep the state, so if trim() is to be used, it must be done in a copy.
uint16_t i = 0;
while (sol_str_slice_split_iterate(content_type, &type, ',')) {
    struct sol_str_slice token = SOL_STR_SLICE_EMPTY;
    struct sol_http_content_type_priority *pri = NULL;

    while (sol_str_slice_split_iterate(type, &token, ';')) {
        struct sol_str_slice clean_token = sol_str_slice_trim(token);

        if (!pri)  {
            pri = sol_vector_append(priorities);
            pri->index = i;
            pri->qvalue = 1.0;
           pri->content_type = clean_token;
           sol_vector_init(&pri->tokens, sizeof(struct sol_str_slice));
           set_type_and_sub_type(pri->content_type, &pri->type, &pi->sub_type);
        } else if (!is_qvalue_start(clean_token)) {
           struct sol_str_slice *el = sol_vector_append(&pri->tokens);
           *el = clean_token;
        } else {
           pri->qvalue = ...;
        }
    }
    i++;
}
```